### DISCUSSION TITLE
CodeViewer - how to auto-format JSON code when pasting or setting content

### DISCUSSION THREAD

**Discussion by benryanwilliams** [GrapesJS/grapesjs#4597] [Q&A]
If I paste code that is already formatted correctly into the `codeViewer` then it is displayed properly. However, if I paste an unformatted block of code into the codeViewer (or set the content of the codeViewer using a stringified JSON object), it just appears as one block of text:

https://user-images.githubusercontent.com/67364267/190609955-69247731-9f61-48bc-bb3f-d760478ec3b3.mov

My code:

```
export default (editor, config) => {

  var setData = function() {console.log("setData function hasn't been replaced")}

  const stylePrefix = editor.getConfig("stylePrefix")
  const modal = editor.Modal
  const codeViewer = editor.CodeManager.getViewer("CodeMirror").clone()
  const container = document.createElement("div")
  const importLabel = config.modalImportLabel
  let viewerEditor = codeViewer.editor

  // Init import button
  const btnSet = document.createElement("button")
  btnSet.innerHTML = config.modalBtnImport || "Set Data"
  btnSet.className = stylePrefix + "btn-prim " + stylePrefix + "btn-import"
  btnSet.onclick = () => {
    if (confirm("Are you sure you want to set the sample data? Existing for-each containers and conditional text blocks may no longer function properly")) {
      let code = codeViewer.editor.getValue()
      code ? setData(JSON.parse(code)) : setData(null)
      editor.Modal.close()
    }
  }

  // Init code viewer
  codeViewer.set({
    codeName: 'htmlmixed',
    readOnly: 0,
    theme: 'hopscotch',
    autoBeautify: true,
    autoCloseTags: true,
    autoCloseBrackets: true,
    lineWrapping: true,
    styleActiveLine: true,
    smartIndent: true,
    indentWithTabs: true
  })

  return {
    run(setThisData, thisData) {
      setData = setThisData

      if (!viewerEditor) {
        const txtarea = document.createElement("textarea")

        if (importLabel) {
          const labelEl = document.createElement("div")
          labelEl.className = `${stylePrefix}import-label`
          labelEl.innerHTML = importLabel
          container.appendChild(labelEl)
        }

        container.appendChild(txtarea)
        container.appendChild(btnSet)
        codeViewer.init(txtarea)
        viewerEditor = codeViewer.editor
      }

      modal.setTitle("Set Sample Data (paste JSON below)")
      modal.setContent(container)
      const content = thisData ? JSON.stringify(thisData) : "Replace this with your sample data"
      codeViewer.setContent(content)
      modal
        .open()
        .getModel()
      viewerEditor.refresh()
    },

    stop() {
      modal.close()
    },
  }
}
```

Even if I use the third parameter of JSON.stringify e.g. `codeViewer.setContent(JSON.stringify(thisData, null, '\t'))`, there is no indentation:

<img width="857" alt="Screenshot 2022-09-16 at 17 10 50" src="https://user-images.githubusercontent.com/67364267/190615965-575b84c3-9ecf-42d2-a1eb-ef299e03bfea.png">

Is there any built-in (or other) way of implementing this? I can't see much information regarding the `codeViewer` in the docs.

**Comment by artf** [comment]
By using the CodeMirror in the core you should be able to auto-format html/css/js by using something like this:
```js
viewerEditor.autoFormatRange({ line:0, ch:0 }, { line: viewerEditor.lineCount() });
```
but for the JSON is generally enough `JSON.stringify`, did you try using simple spaces `JSON.stringify(thisData, null, 2)`?

**Reply by benryanwilliams** [reply]
@artf I've tried adding this line `viewerEditor.autoFormatRange({ line:0, ch:0 }, { line: viewerEditor.lineCount() });` but it doesn't seem to have any effect. Also, if I log `content` then paste it into the codeViewer from the console it appears properly. But then if I set the data then open the dialog again it appears without indentation:


https://user-images.githubusercontent.com/67364267/193527753-6fb5cc06-148e-4fd4-a3e5-a87f0333980f.mov

Also, changing it to `JSON.stringify(thisData, null, 2)` produces the same results. I'm thinking there could be a hacky copy and paste simulation solution but feel like it should be easier to do this - any more ideas?
